### PR TITLE
Give position to symbols when unpickling

### PR DESCRIPTION
**Pull request by smarter**
I started working on this to fix a crash when compiling the collection-strawman contrib tests. I minimized it to `anonBridge.scala` (in the collection-strawman, unpickling happened because of `inline def`, here the test is in `from-tasty` so it will be compiled from an unpickled tree). Eventually, I would also like to use the unpickled symbol positions in the IDE, but this will require more work. (Right now, going from an arbitrary Position to a SourcePosition is annoying, it requires creating a SourceFile which requires converting the source file content to an array of bytes).

### CODE REVIEW

**Comment by smarter** [comment]
test performance please

File: compiler/src/dotty/tools/dotc/config/ScalaSettings.scala [hunk 1]
```diff
@@ -93,6 +93,7 @@ class ScalaSettings extends Settings.SettingGroup {
   val YdetailedStats = BooleanSetting("-Ydetailed-stats", "show detailed internal compiler stats (needs Stats.enabled to be set to true).")
   val Yheartbeat = BooleanSetting("-Ydetailed-stats", "show heartbeat stack trace of compiler operations (needs Stats.enabled to be set to true).")
   val Yprintpos = BooleanSetting("-Yprintpos", "show tree positions.")
+  val YprintposSyms = BooleanSetting("-Yprintpos-syms", "show symbol definitions positions.")
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/config/ScalaSettings.scala]: Since `-Yprintpos` is about to change to `Yprint-pos`, I guess we also need an additional hypen here: `-Yprint-pos-syms`.

File: compiler/src/dotty/tools/dotc/core/Symbols.scala [hunk 2]
```diff
@@ -437,7 +437,7 @@ object Symbols {
 
     /** Does this symbol come from a currently compiled source file? */
     final def isDefinedInCurrentRun(implicit ctx: Context): Boolean = {
-      pos.exists && defRunId == ctx.runId
+      pos.exists && defRunId == ctx.runId && sourceFile == ctx.source.file
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/core/Symbols.scala]: That's not the right condition anymore. See for instance how it is used in `checkNoConflict`. We want to flag top-level definitions in other source files if they clash with a definition in the current source file.

Comment (odersky) [reply compiler/src/dotty/tools/dotc/core/Symbols.scala]: We could probably use as third criterion:

     { val file = assocatedFile
       file != null && !file.path.endsWith(".class")
     }

File: compiler/src/dotty/tools/dotc/parsing/Parsers.scala [hunk 3]
```diff
@@ -1155,7 +1155,8 @@ object Parsers {
         t
     }
 
-    def ascription(t: Tree, location: Location.Value) = atPos(startOffset(t), in.skipToken()) {
+    def ascription(t: Tree, location: Location.Value) = atPos(startOffset(t)) {
+      in.skipToken()
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/parsing/Parsers.scala]: Why the change? The previous implementation would place the `^` under the `:` of an ascription, which seems to be what we want, no?

File: compiler/src/dotty/tools/dotc/transform/ExpandPrivate.scala [hunk 4]
```diff
@@ -90,8 +90,8 @@ class ExpandPrivate extends MiniPhase with IdentityDenotTransformer { thisPhase
       }
 
       assert(d.symbol.sourceFile != null &&
-             isSimilar(d.symbol.sourceFile.path, ctx.source.file.path),
-          s"private ${d.symbol.showLocated} in ${d.symbol.sourceFile} accessed from ${ctx.owner.showLocated} in ${ctx.source.file}")
+             isSimilar(d.symbol.sourceFile.path, ctx.owner.sourceFile.path),
+          s"private ${d.symbol.showLocated} in ${d.symbol.sourceFile} accessed from ${ctx.owner.showLocated} in ${ctx.owner.sourceFile}")
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/transform/ExpandPrivate.scala]: When would this make a difference?

File: compiler/src/dotty/tools/dotc/config/ScalaSettings.scala (hunk 1 above)
Comment (smarter) [reply compiler/src/dotty/tools/dotc/config/ScalaSettings.scala]: Sure, in which PR is Yprintpos changed to Yprint-pos ?

File: compiler/src/dotty/tools/dotc/core/Symbols.scala (hunk 2 above)
Comment (smarter) [reply compiler/src/dotty/tools/dotc/core/Symbols.scala]: Hmm that's tricky because associatedFile forces things, I'll give it a try.

File: compiler/src/dotty/tools/dotc/parsing/Parsers.scala (hunk 3 above)
Comment (smarter) [reply compiler/src/dotty/tools/dotc/parsing/Parsers.scala]: In all other cases, the point is under the first letter of the name of the definition (e.g. for `def foo: Int = 1` the point is under `f`). This change means that I can always use the point of the definition tree to set the symbol position.

File: compiler/src/dotty/tools/dotc/transform/ExpandPrivate.scala (hunk 4 above)
Comment (smarter) [reply compiler/src/dotty/tools/dotc/transform/ExpandPrivate.scala]: When unpickling from tasty, `ctx.source.file` will be "Foo.class" whereas `ctx.owner.sourceFile` will be the source file name found in the `SourceFile` annotation of the tasty tree.

**Comment by smarter** [comment]
@odersky There's another difficulty with `isDefinedInCurrentRun` If we're compiling from TASTY then `associatedFile` may very well end with `.class` but `isDefinedInCurrentRun` should return true. The only solution I see here is doing a check like `ctx.runInfo.units.contains(associatedFile)`, but that requires adding `units` to `RunInfo` (currently it's only in `Run` which isn't accessible from `Context`), what do you think?

**Comment by smarter** [comment]
Alternative proposal to make `isDefinedInCurrentRun` easy to define: track whether a symbol comes from a compilation unit when the symbol is created. This could be done by giving a special meaning to one bit of the `Coord` class

**Comment by smarter** [comment]
test performance please

**Comment by smarter** [comment]
@odersky Could you review the last commit please?

**Comment by smarter** [comment]
Ping for review of the last commit

**Comment by odersky** [comment]
I am worried that this seems to cause a slowdown of 6% for Dotty itself. What could be the cause of this?

**Comment by odersky** [comment]
Last commit LGTM, but it would be good to drill down on the performance issue.

**Comment by smarter** [comment]
test performance please

**Comment by odersky** [comment]
I am still not convinced this is the right thing to do. We add a lot more positions to Tasty files which makes them larger. Symbols get larger too. It seems this has a significant performance implications. So what exactly are we trying to solve here? Can't this be obtained more cheaply? 

If it's just the inlining, it would be easy to fix with an additional pass when we create the Inline Body annotation.

**Comment by smarter** [comment]
> We add a lot more positions to Tasty files which makes them larger.

Not at all! The positions of symbols are derived from the position of trees, so we're not storing anything more.

>  Symbols get larger too.

I don't see how, symbols already had a `coord` field. It's just that for many of them it was set to `NoCoord` previously.

>  It seems this has a significant performance implications.

I think the benchmark result was a fluke, we'll see. If there are performance issues I'll fix them.

> So what exactly are we trying to solve here?

Unpickled trees should be as close to their pre-pickling state as possible. This reduces the number of surprises when compiling from tasty since we're using the same code paths than when compiling from sources. Positions for symbols may also allow for better error messages, and could be used in the IDE to reduce the number of tree traversals done for operations such as jump-to-definition (the current PR doesn't do this because it's not trivial).

**Comment by smarter** [comment]
Hmm it looks like we have no new benchmarks recorded for master since 20/11 (see http://dotty-bench.epfl.ch), so it's difficult to measure the effect of this PR. Ping @liufengyun

**Comment by smarter** [comment]
test performance please: f92f278 6d5c226

**Comment by smarter** [comment]
Oops, used the wrong syntax to test multiple commits, trying again...

**Comment by smarter** [comment]
test performance please 7848ef6 6b19932

**Comment by smarter** [comment]
@liufengyun Looks like testing multiple commits doesn't work anymore, or did I get the syntax wrong?

**Comment by liufengyun** [comment]
You need a comma after the please.

Le 6 déc. 2017 19:43, "Guillaume Martres" <notifications@github.com> a
écrit :

**Comment by odersky** [comment]
OK, I see the rationale now, and it seems performance implications are smaller than first feared. So I am OK to get this in. But maybe finish the performance evaluation first.

**Comment by smarter** [comment]
test performance please; 18767e0d 6b19932

**Comment by liufengyun** [comment]
@smarter Sorry, it's colon instead semicolon, I've a little problem with the punctuation names.

**Comment by smarter** [comment]
I tried using colon in https://github.com/lampepfl/dotty/pull/3525#issuecomment-349654481 but it didn't seem to work either ?

**Comment by liufengyun** [comment]
Let's try a new curve and see:

retest performance please: 18767e0d 6b19932

**Comment by liufengyun** [comment]
retest performance please: 18767e0d 6b19932
